### PR TITLE
Fetch from AD in pages

### PR DESCRIPTION
**Pull request by MrManor**
This allows for more than 1000 results from AD, as long as you do not run out of php memory.

### CODE REVIEW

**Comment by piRGoif** [comment]
This follow discussion made in [SF#1768](https://sourceforge.net/p/itop/tickets/1768/)

File: LDAPCollector.class.inc.php [hunk 1]
```diff
@@ -154,15 +154,40 @@ public function Search($sDN, $sFilter, $aAttributes = array('*'))
         if ($this->Connect())
         {
             Utils::Log(LOG_DEBUG, "ldap_search('$sDN', '$sFilter', ['".implode("', '", $aAttributes)."'])...");
+
+	$cookie = '';
+	$pageSize = 500;
+	$firstPass = true;
+	do {
+	    ldap_control_paged_result($this->rConnection, $pageSize, true, $cookie);
```
Comment (piRGoif) [inline LDAPCollector.class.inc.php:162]: This is deprecated :/
See https://www.php.net/manual/en/function.ldap-control-paged-result : 
> This function has been DEPRECATED as of PHP 7.4.0, and REMOVED as of PHP 8.0.0. Instead the controls parameter of ldap_search() should be used. See also LDAP Controls for details.

File: LDAPCollector.class.inc.php [hunk 2]
```diff
@@ -154,15 +154,40 @@ public function Search($sDN, $sFilter, $aAttributes = array('*'))
...
+	    ldap_control_paged_result_response($this->rConnection, $rSearch, $cookie);
+	    $iNumberUser = count($aList) - 1;
+	    Utils::Log(LOG_INFO,"(Persons) Number of entries found on LDAP so far: ".$iNumberUser);
+	    $firstPass = false;
+	} while($cookie !== null && $cookie != ''); 		
+
+
             
-            $aList = ldap_get_entries($this->rConnection, $rSearch);
+	    //$aList = ldap_get_entries($this->rConnection, $rSearch);
```
Comment (piRGoif) [inline LDAPCollector.class.inc.php:190]: You can remove the line, your code replace it, and it his versioned by git.

File: LDAPCollector.class.inc.php [hunk 3]
```diff
@@ -154,15 +154,40 @@ public function Search($sDN, $sFilter, $aAttributes = array('*'))
...
+	    if($firstPass)
+	    {
+		$aList = $bList;
+	    }
+	    else
+	    {
+		$aList = array_merge($aList, $bList);
+	    }
+
+	    ldap_control_paged_result_response($this->rConnection, $rSearch, $cookie);
```
Comment (piRGoif) [inline LDAPCollector.class.inc.php:182]: Same remark, this is deprecated
https://www.php.net/manual/en/function.ldap-control-paged-result-response

**Review by piRGoif — CHANGES_REQUESTED** [review]
@dflaven reviewed your code, I'm reporting his comments.

First, congratulations, you did find the right place to modify !

But... your code uses deprecated methods (in 7.4.0), so we can't accept it right now :(
Can you propose a new version using maintained API ?

Also, note that we'll have to be very careful as it seems that not all servers support pagination. So we might need to add a feature flag, and by default retrieve the old way, and for people knowing their server support the pagination, do it the new way ?

**Comment by MrManor** [comment]
As discussed @ sourceforge am Im not a programmer and I was reluctant to make this pr in the first place. The hack was done by a co-worker, - and I think more work should be done to sync download and upload to conserve php memory.

**Comment by piRGoif** [comment]
Hello @MrManor,
Yes you're right you told us this on SF, sorry  :/
I'm reopening this PR as it is a good starting point ! Maybe someone in Combodo or in the community will be able to improve it O:)

**Review by Hipska — APPROVED** [review]
Since this collector only supports (according to documentation) php 5.3 to 7.2 and iTop latest release supports php up to 7.4 and the paged method only has deprecation status since 7.4, I see no problem to merge anyway.

The code can still be refactored when this extension or iTop starts supporting php 8, and I am happy to help with that.

**Comment by piRGoif** [comment]
Ok, the recommended API (LDAP controls) was added in PHP 7.3, I missed this information :/ 
As for now PHP 7.3 is only in security support, not sure this would be a good idea for us to implement something as is.

Hipska if you've got a solution with the new API it would be much welcomed !

**Comment by Hipska** [comment]
Now you confuse me? Do you still think the new API must be implemented instead of the solution in this PR?

**Comment by piRGoif** [comment]
Sorry, my previous comment was quite ambigous
I meant I would still prefer implementing a solution with the PHP 7.3+ API, even if it means to raise this extension's PHP requirements by a lot : PHP 7.3 should now be the very minimum version to be used...
Maybe my opinion isn't shared by other Combodo employees

**Comment by Hipska** [comment]
I would not go that way, since minimal php version for iTop is 5.6 and for the upcoming 3.0 it is even 7.1.3.

**Comment by piRGoif** [comment]
Indeed for now I don't think we already had extensions that had requirements stronger than iTop...
We will discuss this internally...

**Comment by piRGoif** [comment]
This problem is referenced in the (private) Combodo bug DB as N°2210. It has already been planned for the next release of the collector. Having this code will be a very good start !

**Comment by piRGoif** [comment]
Hello,
The code was updated with a different solution, so I'm closing this PR. 
A new version should be published on the hub by the end of the month, but modifications are directly available [in itop-data-collector-base repo](https://github.com/Combodo/itop-data-collector-ldap/commit/63818f2cdb5bafe36212757f1fe0d62c1be72452)

**Pull request closed without merging**